external-crates/move/crates/move-package/src/compilation/versioned_compilation.rs

```diff
@@ -0,0 +1,196 @@
+use crate::{
+    lock_file::{self, schema::ToolchainVersion},
+    source_package::layout::SourcePackageLayout,
+};
+use anyhow::{anyhow, bail, Result};
+use colored::Colorize;
+use move_command_line_common::env::MOVE_HOME;
+use move_compiler::shared::PackagePaths;
+use move_symbol_pool::Symbol;
+
+use std::process::Command;
+use std::{
+    ffi::OsStr,
+    fs::File,
+    io::Write,
+    path::{Path, PathBuf},
+};
+use tracing::debug;
+
+/// partitions `deps` by whether we need to compile dependent packages with a
+/// prior toolchain (which we find by looking at Move.lock contents) or
+/// whether we can compile them with the current binary.
+pub fn partition_deps_by_toolchain<W: Write>(
+    deps: Vec<PackagePaths>,
+    current_compiler_version: Option<String>,
+    w: &mut W,
+) -> Result<(Vec<PackagePaths>, Vec<PackagePaths>)> {
+    let current_compiler_version = current_compiler_version.unwrap_or_else(|| "0.0.0".into());
+    debug!("current compiler: {current_compiler_version}");
+    let mut deps_for_current_compiler = vec![];
+    let mut deps_for_prior_compiler = vec![];
+    for dep in deps {
+        let a_source_path = dep.paths[0].as_str();
+        let root = SourcePackageLayout::try_find_root(Path::new(a_source_path))?;
+        let lock_file = root.join("Move.lock");
+        if !lock_file.exists() {
+            deps_for_current_compiler.push(dep);
+            continue;
+        }
+
+        let mut lock_file = File::open(lock_file)?;
+        let toolchain_version = lock_file::schema::ToolchainVersion::read(&mut lock_file)?;
+        match toolchain_version {
+            // No ToolchainVersion implies current compiler
+            None => deps_for_current_compiler.push(dep),
```

Comment:
This case needs to be split into two, which I will do in follow up:

- `Move.lock` version does **not** know about `toolchain-version` => legacy compiler
- `Move.lock` version knows about toolchain-version but doesn't specify it => current compiler